// @storybook/addon-actions/dist/preview.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _toConsumableArray2 = require('babel-runtime/helpers/toConsumableArray');

var _toConsumableArray3 = _interopRequireDefault(_toConsumableArray2);

var _from = require('babel-runtime/core-js/array/from');

var _from2 = _interopRequireDefault(_from);

exports.action = action;
exports.decorateAction = decorateAction;

var _addons = require('@storybook/addons');

var _addons2 = _interopRequireDefault(_addons);

var _jsonStringifySafe = require('json-stringify-safe');

var _jsonStringifySafe2 = _interopRequireDefault(_jsonStringifySafe);

var _ = require('./');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _format(arg) {
  if (arg && typeof arg.preventDefault !== 'undefined') {
    return (0, _jsonStringifySafe2.default)('[SyntheticEvent]');
  }
  return (0, _jsonStringifySafe2.default)(arg);
}

function action(name) {
  var handler = function handler() {
    for (var _len = arguments.length, _args = Array(_len), _key = 0; _key < _len; _key++) {
      _args[_key] = arguments[_key];
    }

    var args = (0, _from2.default)(_args).map(_format);
    var channel = _addons2.default.getChannel();
    var randomId = Math.random().toString(16).slice(2);
    channel.emit(_.EVENT_ID, {
      id: randomId,
      data: { name: name, args: args }
    });
  };

  // some day when {[name]: function() {}} syntax is not transpiled by babel
  // we can get rid of this eval as by ES2015 spec the above function gets the
  // name `name`, but babel transpiles to Object.defineProperty which doesn't do
  // the same.
  //
  // Ref: https://bocoup.com/weblog/whats-in-a-function-name
  var fnName = name ? name.replace(/\W+/g, '_') : 'action';
  var named = eval('(function ' + fnName + '() { return handler.apply(this, arguments) })');
  return named;
}

function decorateAction(decorators) {
  return function (name) {
    var callAction = action(name);
    return function () {
      for (var _len2 = arguments.length, _args = Array(_len2), _key2 = 0; _key2 < _len2; _key2++) {
        _args[_key2] = arguments[_key2];
      }

      var decorated = decorators.reduce(function (args, fn) {
        return fn(args);
      }, _args);
      callAction.apply(undefined, (0, _toConsumableArray3.default)(decorated));
    };
  };
}
